libompd/src/TargetValue.cpp

```diff
@@ -349,6 +349,60 @@ ompd_rc_t TValue::getRawValue(void *buf, int count) {
   return errorState.errorCode;
 }
 
+ompd_rc_t TValue::getString(const char **buf) {
+  if (gotError())
+    return getError();
+
+  TValue strValue = dereference();
+  if (strValue.gotError() && (strValue.getError() != ompd_rc_unsupported)) {
+    return strValue.getError();
+  }
+
+  ompd_rc_t ret;
+
+  if (strValue.gotError() && (strValue.getError() == ompd_rc_unsupported)) {
+    if (strValue.symbolAddr.address == 0) {
+      // Pass back an empty string if the pointer is NULL.
+      char *empty_string;
+      ret = callbacks->alloc_memory(1, (void **)&empty_string);
+      if (ret != ompd_rc_ok) {
+        return ret;
+      }
+      empty_string[0] = '\0';
+      *buf = empty_string;
+      return ompd_rc_ok;
+    }
+    else {
+      return strValue.getError();
+    }
+  }
+  // Determine the length of the string needed by reading in 1 character
```

Comment:
This seems like a very expensive way to read the string. If you don't know the approximate length of the string, why not read it in chunks? For example, read the string 512 bytes at a time until the strlen() of the string that is returned by the callback is less than 512, or you run into a read error. It's probably much, much cheaper to loop calling alloc_memory() and read_string() on modest sized chunks than it is to read one character at a time. Note that accessing the target's memory might be very expensive, especially for debugger client/server setups.

Reply:
I think, the reasoning is that from `read_string()` the OMP library cannot tell, whether the complete string was fetched.
I agree, that we should assume the latency of debugee lookups (via callbacks) to be a bottleneck.
In any case, a local search for `\0` in the provided buffer should be more efficient than repeated single-byte reads.

Reply:
Makes sense. I was not quite comfortable with the expense too, but as @jprotze mentioned, I had implemented it this way since there was no straightforward way to tell if the complete string was fetched. Anyways, I am making changes to call read_string() with a size of 512 bytes and return an error on the failure. I am currently having the read_string() callback routine in the debugger determine if the string is truncated (using a local search for \0) and return an error (which later needs to be ompd_rc_incomplete). This makes me think that shouldn't a callback realloc_memory() routine also be provided by the tool along with the alloc_memory() and free_memory()  routines ? It might be a cleaner approach.

Reply:
> I am currently having the read_string() callback routine in the debugger determine if the string is truncated (using a local search for \0) and return an error (which later needs to be ompd_rc_incomplete). 

I think this is the **wrong** thing to do. The semantics of read_string() are specified in the OMP v5 standard, so unless we change the standard, the implementation of read_string() in the debugger (I assume you mean GDB) should follow the standard. The standard specifies what happens if the string does not contain a null, which is to return success and nbytes of the string.

WRT realloc_memory(), as far as I know, there is no restriction in OMPD that prevents you from calling realloc() internally, as long as you don't pass that pointer back to the debugger. You could build up the string internally in a realloc'd buffer, which gives you the length, and then you can use the alloc_memory() callback to allocate a string, strcpy(), and to pass back to the debugger.

Reply:
Really sorry, John, but could you point me to the section in the standard that mentions that if the string does not contain a null, read_string() should return success and nbytes ? I see this in the standard:

"At most nbytes bytes are copied. If a null byte is not among the first
nbytes bytes, the string placed in buffer is not null-terminated." 

But I don't see anything regarding a return value. 

Regarding realloc(), I thought the standard mentions that OMPD should always use the tool for its memory management. From Section 5.4.1., "Memory Management of OMPD Library".

=========snippet from the standard ====================
The OMPD library must not access the heap manager directly. Instead, if it needs heap memory it
must use the memory allocation and deallocation callback functions that are described in this
section, ompd_callback_memory_alloc_fn_t (see Section 5.4.1.1 on page 546) and
ompd_callback_memory_free_fn_t (see Section 5.4.1.2 on page 546), which are provided
by the tool to obtain and to release heap memory. This mechanism ensures that the library does not
interfere with any custom memory management scheme that the tool may use.
===========================================

Reply:
Sorry for the font !